images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go

```diff
@@ -0,0 +1,160 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch {
+	case lease == nil && vmipStatus.Address != "":
+		h.logger.Info("Lease by name not found: waiting for the lease to be available")
+		return reconcile.Result{}, nil
+
+	case lease == nil:
+		h.logger.Info("No Lease for VirtualMachineIP: create the new one", "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+		return h.createNewLease(ctx, state)
+
+	case lease.Status.Phase == "":
+		// TODO: Replace this requeue with proper UpdateFunc in VirtualMachineIPAddressLease watch setup.
+		h.logger.Info("Lease is not ready: waiting for the lease")
+		return reconcile.Result{Requeue: true, RequeueAfter: 2 * time.Second}, nil
+
+	case util.IsBoundLease(lease, vmip):
+		h.logger.Info("Lease already exists, VirtualMachineIP ref is valid")
+		return reconcile.Result{}, nil
+
+	case lease.Status.Phase == virtv2.VirtualMachineIPAddressLeasePhaseBound:
+		h.logger.Info("Lease is bounded to another VirtualMachineIP: recreate VirtualMachineIP when the lease is released")
+		return reconcile.Result{}, nil
+
+	default:
+		h.logger.Info("Lease is released: set binding")
+
+		if lease.Spec.VirtualMachineIPAddressRef.Namespace != vmip.Namespace {
+			return reconcile.Result{}, fmt.Errorf("the selected VirtualMachineIP lease belongs to a different namespace: %s", lease.Spec.VirtualMachineIPAddressRef.Namespace)
+		}
+
+		lease.Spec.VirtualMachineIPAddressRef = &virtv2.VirtualMachineIPAddressLeaseIpAddressRef{
+			Name:      vmip.Name,
+			Namespace: vmip.Namespace,
+		}
+
+		err := h.client.Update(ctx, lease)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		vmipStatus.Address = common.LeaseNameToIP(lease.Name)
+		return reconcile.Result{}, nil
+	}
+}
+
+func (h IPLeaseHandler) createNewLease(ctx context.Context, state state.VMIPState) (reconcile.Result, error) {
+	vmip := state.VirtualMachineIP()
+	vmipStatus := &vmip.Status
+
+	if vmip.Spec.Type == virtv2.VirtualMachineIPAddressTypeAuto {
+		h.logger.Info("allocate the new VirtualMachineIP address")
+		var err error
+		vmipStatus.Address, err = h.ipService.AllocateNewIP(state.AllocatedIPs())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		vmipStatus.Address = vmip.Spec.StaticIP
+	}
+
+	if !h.ipService.IsAvailableAddress(vmipStatus.Address, state.AllocatedIPs()) {
+		msg := fmt.Sprintf("VirtualMachineIP cannot be created: the address %s has already been allocated for another vmip", vmip.Spec.StaticIP)
+		h.logger.Info(msg)
+		h.recorder.Event(vmip, corev1.EventTypeWarning, vmipcondition.VirtualMachineIPAddressLeaseLost, msg)
+		return reconcile.Result{}, nil
+	}
+
+	leaseName := common.IpToLeaseName(vmipStatus.Address)
+
+	h.logger.Info("Create lease",
+		"leaseName", leaseName,
+		"refName", vmip.Name,
+		"refNamespace", vmip.Namespace,
+	)
+
+	err := h.client.Create(ctx, &virtv2.VirtualMachineIPAddressLease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: leaseName,
```

Comment:
Add controller owner ref

Reply:
Adding an owner ref will delete the resource VirtualMachineIPAddressLease after the resource VirtualMachineIPAddress, and by design, the resource VirtualMachineIPAddressLease monitors its own deletion.